chore(docs): Created `GitHub runner cryptographic overview` (#386)
canonical/github-runner-operator: docs/reference/cryptographic-overview.md

# GitHub runner cryptographic overview
This document provides an overview of the cryptographic technologies used in the GitHub runner charm, including encryption/decryption, hashing, and digital signatures.

> ## Overall Description of Cryptographic Technology Use

Reviewer: In general, simpler grammar is preferred for technical documentation to lower the barrier for comprehension, for instance, some suggestions
s/that need to be executed/to be executed
s/retrieve information about how the repositories are set up/retrieve the repositories' configuration
s/the runner creates virtual machines in OpenStack and uses these machines to run the workload/the runner creates virtual machines in OpenStack to run the workload

Would you mind double checking if you can further simplify?

Author: Sure can do! Thanks for the tip!

> The GitHub runner charm uses various cryptographic technologies to ensure secure communication and data integrity. The following sections describe each of the components.

Reviewer 2: Should we have a new line to separate the headers? :D

### TLS
The GitHub runner communicates with GitHub to receive information about the workflows to be executed, retrieve the repositories config and send back workflow logs to the user. This communication uses [urllib3](https://urllib3.readthedocs.io/en/stable/) under the hood using TLS 1.3.

> In OpenStack mode, the runner creates virtual machines to run the workload in OpenStack. The GitHub runner charm interacts with the OpenStack API to create and manage virtual machines. These interactions are secured via TLS. The communication between these virtual machines and runner is done via SSH and secured by 256 byte keypairs.

Reviewer 2: Nit: would the following info be helpful?

The SSH key is created by OpenStack, which uses ssh PEM keys (OpenSSL).
```
$ ssh-keygen -l -f ./<keyfile>
2048 SHA256:w0gf4CC2xY3X9z+WRwT1xh4osy1j25hsm3fCg7bUOqk no comment (RSA)
```

> [DockerHub charm](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).

Reviewer 3: ```suggestion
[Docker Registry](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
```

> Images that run in the OpenStack vm are built using the [Image Builder Charm](https://github.com/canonical/github-runner-image-builder-operator). This charm needs to download the runner binary, cloud image to base the image on and yq. All these images are downloaded with TLS.

Reviewer 3: nit: Technically its not the charm, but the application managed by the charm, so maybe rephrase to "This charm manages an application that needs to download the runner binary..." (cc @yanksyoon )

The GitHub runner charm supports being deployed behind an HTTP proxy. [Aproxy](https://github.com/canonical/aproxy) is installed and enabled when an HTTP proxy is detected so that jobs executing in the runner VMs don’t have to configure the proxy themselves. Aproxy is a transparent proxy service for HTTP and HTTPS/TLS connections. Aproxy works by pre-reading the Host header in HTTP requests and SNI in TLS hellos; it forwards HTTP proxy requests with the hostname therefore, complies with HTTP proxies requiring destination hostname for auditing or access control. Aproxy doesn't and can't decrypt the TLS connections. It works by reading the plaintext SNI information in the client hello during the TLS handshake, so the authentication and encryption of TLS are never compromised. Aproxy supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.
### Signature Verification

> Cloud image is verified by SHA256 checksum. Runner binary is also downloaded by [GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.

Reviewer 3: Could we elaborate which Cloud image is meant here?

Reviewer 3: I think this sentence should be updated here, too.

### User SSH Access

> Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.

Reviewer 3: ```suggestion
Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses the SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.
```

## Cryptographic Technology Used by the Product
The following cryptographic technologies are used internally by our product:
### TLS
- Communication with GitHub is done via TLS v1.3
- [The Repo Policy Compliance tool](https://github.com/canonical/repo-policy-compliance) communicates with the [GitHub API](https://docs.github.com/en/rest?apiVersion=2022-11-28) using TLS v1.3.
- Communication with the OpenStack API uses TLS v1.3.
### Signature Verification
Cloud images are verified by SHA256 checksum. Runner binary is also downloaded by [the GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
### User SSH Access
Tmate secures SSH connections using [OpenSSL](https://www.openssl.org/).
### RSA

> The communication between these virtual machines and runner is done via SSH and secured by RSA 256 byte keypairs.

Reviewer 3: ```suggestion
The communication between these virtual machines and rthe unner is done via SSH and secured by RSA 256 byte keypairs.
```

## Cryptographic Technology Exposed to Users
The following sections describe the cryptographic technologies exposed to the user:
### Tmate

> - [Tmate](https://tmate.io/) uses RSA (384 byte) and ED25519 (32 byte) fingerprints for connections from users to [Tmate](https://tmate.io/) and from the managers to [Tmate](https://tmate.io/). [OpenSSL](https://www.openssl.org/) is being used by [Tmate](https://tmate.io/) to secure the connection between user/manager to runner.

Reviewer 3: ```suggestion
- [Tmate](https://tmate.io/) uses RSA (384 byte) and ED25519 (32 byte) fingerprints for connections from users to [Tmate](https://tmate.io/) and from the managers to [Tmate](https://tmate.io/). [OpenSSL](https://www.openssl.org/) is being used by [Tmate](https://tmate.io/) to secure the connection between the user/manager to the runner.
```

### Docker Hub Cache
[Docker Hub cache](https://github.com/canonical/docker-registry-charm) connection is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
### Aproxy
[Aproxy](https://github.com/canonical/aproxy) supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.
## Packages or Technology Providing Cryptographic Functionality
The following packages or technologies provide cryptographic functionality:
- [OpenSSL](https://www.openssl.org/) library is being used.
- For TLS and HTTPS connections urllib3 python library is used.
- Default clients in Ubuntu for SSH and TLS are being used.

> - [Python hashlib](https://docs.python.org/3/library/hashlib.html) package is being used for sha256 checksum calculation/verification of runner binary and Cloud Init.

Reviewer 4: Nitpick: Does sha256 need to be capitalised?

Author: When I search for it online it is always capitalized but still wrong I guess. I see it as `SHA-256` in a lot of places.

Ex.: [Wikipedia](https://en.wikipedia.org/wiki/SHA-2), [SimpliLearn](https://www.simplilearn.com/tutorials/cyber-security-tutorial/sha-256-algorithm)

# GitHub runner cryptographic overview

> This document provides an overview of the cryptographic technologies used in the GitHub runner charm, including encryption/decryption, hashing, and digital signatures.

Reviewer 4: Two nitpicks on this sentence: 
1. I think you can separate encryption and decryption with a comma 
2. The style guide recommends not using the Oxford comma, so remove the comma after "hashing"

The GitHub runner charm supports being deployed behind an HTTP proxy. [Aproxy](https://github.com/canonical/aproxy) is installed and enabled when an HTTP proxy is detected so that jobs executing in the runner VMs don’t have to configure the proxy themselves. Aproxy is a transparent proxy service for HTTP and HTTPS/TLS connections. Aproxy works by pre-reading the Host header in HTTP requests and SNI in TLS hellos; it forwards HTTP proxy requests with the hostname therefore, complies with HTTP proxies requiring destination hostname for auditing or access control. Aproxy doesn't and can't decrypt the TLS connections. It works by reading the plaintext SNI information in the client hello during the TLS handshake, so the authentication and encryption of TLS are never compromised. Aproxy supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.

### Signature Verification
Images that run in the OpenStack vm (Cloud images) are verified by SHA256 checksum. Runner binary is also downloaded by [GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.

### User SSH Access

> Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses the SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.

Reviewer 4: Nitpick: Capitalise RSA (I see it's capitalised elsewhere in the document, so for consistency it should be capitalised here)

### RSA
The communication between these virtual machines and the runner is done via SSH and secured by RSA 256 byte keypairs.

## Cryptographic Technology Exposed to Users
The following sections describe the cryptographic technologies exposed to the user:

### Tmate
- [Tmate](https://tmate.io/) uses RSA (384 byte) and ED25519 (32 byte) fingerprints for connections from users to [Tmate](https://tmate.io/) and from the managers to [Tmate](https://tmate.io/). [OpenSSL](https://www.openssl.org/) is being used by [Tmate](https://tmate.io/) to secure the connection between the user/manager to the runner.

### Docker Hub Cache
[Docker Hub cache](https://github.com/canonical/docker-registry-charm) connection is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).

### Aproxy
[Aproxy](https://github.com/canonical/aproxy) supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.

## Packages or Technology Providing Cryptographic Functionality
The following packages or technologies provide cryptographic functionality:
- [OpenSSL](https://www.openssl.org/) library is being used.
- For TLS and HTTPS connections urllib3 python library is used.

> - Default clients in Ubuntu for SSH and TLS are being used.

Reviewer 4: Hi, is it possible to give a brief explanation for what these libraries are being used for?

- [Python hashlib](https://docs.python.org/3/library/hashlib.html) package is being used for SHA256 checksum calculation/verification of runner binary and Cloud Init.
- OpenStack client is being used to generate keypairs.

> - [Aproxy](https://github.com/canonical/aproxy) using the [golang.org/x/crypto](http://golang.org/x/crypto) package.

Reviewer 4: Is it possible to give a brief explanation for what Aproxy is being used for?

## Overall Description of Cryptographic Technology Use
The GitHub runner charm uses various cryptographic technologies to ensure secure communication and data integrity. The following sections describe each of the components.

### TLS
The GitHub runner communicates with GitHub to receive information about the workflows to be executed, retrieve the repositories config and send back workflow logs to the user. This communication uses [urllib3](https://urllib3.readthedocs.io/en/stable/) under the hood using TLS 1.3.

In OpenStack mode, the runner creates virtual machines to run the workload in OpenStack. The GitHub runner charm interacts with the OpenStack API to create and manage virtual machines. These interactions are secured via TLS. The communication between these virtual machines and runner is done via SSH and secured by 256 byte keypairs.

[DockerHub Registry](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).

> Images that run in the OpenStack vm are built using the [Image Builder](https://github.com/canonical/github-runner-image-builder-operator). This application needs to download the runner binary, yq and the cloud image to base the image on. All these images are downloaded with TLS.

Reviewer 3: ```suggestion
Images that run in the OpenStack vm are built using the [Image Builder](https://github.com/canonical/github-runner-image-builder. This application needs to download the runner binary, yq and the cloud image to base the image on. All these images are downloaded with TLS.
```

### Signature Verification
Cloud image is verified by SHA256 checksum. Runner binary is also downloaded by [GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.

### User SSH Access
Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses the SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.

## Cryptographic Technology Used by the Product
The following cryptographic technologies are used internally by our product:

### TLS
- Communication with GitHub is done via TLS v1.3
- [The Repo Policy Compliance tool](https://github.com/canonical/repo-policy-compliance) communicates with the [GitHub API](https://docs.github.com/en/rest?apiVersion=2022-11-28) using TLS v1.3.
- Communication with the OpenStack API uses TLS v1.3.

### Signature Verification

> Cloud images that are downloaded by the Image Builder are verified by SHA256 checksum. Runner binary is also downloaded by [the GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.

Reviewer 3: @yanksyoon Can you have a look on "Cloud images that are downloaded by the Image Builder are verified by SHA256 checksum." if its correct in terms of the technical implementation?

Author: @yanksyoon Is this technically correct?

Reviewer 2: Yup :)
https://github.com/canonical/github-runner-image-builder/blob/main/src/github_runner_image_builder/cloud_image.py#L153

Author: Thanks for confirming!
